### PR TITLE
Ignore MSI violations with zero mutations

### PR DESCRIPTION
**Pull request by Landerstraeten**
- [x] Covered by tests
- [x] Doc PR https://github.com/infection/site/pull/22

We have a [task](https://github.com/phpro/grumphp/blob/master/doc/tasks/infection.md) for Infection in [Grumphp](https://github.com/phpro/grumphp). The Grumphp task uses the `--filter` option of infection so I only listens on changed files.

When Infection runs with a required MSI percentage and there are no mutations generated, Infection results in a failure since the required percentage is not reached.

This PR fast-exits Infection when no mutations are generated.

```
Running initial test suite...

PhpSpec version: 3.4.3



Generate mutants...



.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out



0 mutations were generated:
       0 mutants were killed
       0 mutants were not covered by tests
       0 covered mutants were not detected
       0 errors were encountered
       0 time outs were encountered

Metrics:
         Mutation Score Indicator (MSI): 0%
         Mutation Code Coverage: 0%
         Covered Code MSI: 0%

Please note that some mutants will inevitably be harmless (i.e. false positives).


 [ERROR] The minimum required MSI percentage should be 100%, but actual is 0%.
         Improve your tests!



Creating mutated files and processes:    0/0
```

### CODE REVIEW

**Comment by BackEndTea** [comment]
👍 For the idea and the PR.

I"d like to see a nice message to the user indicating that it exited without generating any mutations.

Also i'm not 100% on giving it a exit code of 0. What do you think @borNfreee

File: src/Command/InfectionCommand.php [hunk 1]
```diff
@@ -200,6 +200,10 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         $mutations = $mutationsGenerator->generate($input->getOption('only-covered'), $input->getOption('filter'));
 
+        if (!((bool) \count($mutations))) {
```
Comment (borNfreee) [inline src/Command/InfectionCommand.php]: `\count($mutations) === 0` if more general, without casting

**Comment by borNfreee** [comment]
> Also i'm not 100% on giving it a exit code of 0.

This is what @Landerstraeten wants to achieve, since now Infection exits with non zero exit code.

my thoughts are:

it's better to have a parameter for this behaviour. Thus, it won't be a BC break. 

I saw different variations in libs about this case, like `--non-zero-exit-on-violation` on phpmnd (magic number detector), `--ignore-violations-on-exit` on PHPMD, etc.

So, we can add an option for such case, where _we want to ignore MSI violations with zero mutations_. I'm not sure about naming though

**Comment by Landerstraeten** [comment]
I vote for `--ignore-msi-with-no-mutations` 😁

**Comment by borNfreee** [comment]
then, could you please

* Create a documentation for this new option here https://github.com/infection/site/blob/master/src/guide/command-line-options.md
* Add a e2e test to make sure it works as expected (you can use https://github.com/infection/infection/blob/master/tests/add_new_e2e)

**Comment by BackEndTea** [comment]
The travis error is due to this branch not being properly based on master. If it is rebased it should have the correct MSI etc.

I'm okay with merging this without an e2e test, as long as it gives some kind of notification to the user that there were 0 mutations, and that it exited succesfully because of that.

**Comment by borNfreee** [comment]
> I'm okay with merging this without an e2e test

Sorry, no new functionality without tests. Just want to minimize sleepless nights after release dates 🤣 

I've added e2e test and simplified the code.

> as long as it gives some kind of notification to the user that there were 0 mutations

It shows the whole Metrics section to the console.

Ready for review.

**Comment by borNfreee** [comment]
Thanks everyone!

**Pull request merged**